src/safir/redis.py

```diff
@@ -0,0 +1,235 @@
+"""Redis database support."""
+
+from __future__ import annotations
+
+from collections.abc import AsyncIterator
+from typing import Generic, Optional, TypeVar
+
+import redis.asyncio as redis
+from cryptography.fernet import Fernet
+from pydantic import BaseModel
+
+from .slack.blockkit import SlackException, SlackMessage, SlackTextField
+
+__all__ = [
+    "PydanticRedisStorage",
+    "EncryptedPydanticRedisStorage",
+    "S",
+    "DeserializeError",
+]
+
+
+class DeserializeError(SlackException):
+    """Raised when a stored Pydantic object in Redis cannot be decoded (and
+    possibly decrypted) or deserialized.
+    """
+
+    def __init__(self, msg: str, key: str) -> None:
+        super().__init__(msg)
+        self.key = key
+
+    def to_slack(self) -> SlackMessage:
+        message = super().to_slack()
+        message.fields.append(SlackTextField(heading="Key", text=self.key))
+        return message
```

Comment:
Because this is Pydantic, which uses newlines and indentation in its exception messages that will be formatted oddly in the main part of the Slack message, I think this would read better in Slack if you did something like this instead:
```suggestion
    def __init__(self, msg: str, key: str, error: str) -> None:
        super().__init__(msg)
        self.key = key
        self.error = error

    def to_slack(self) -> SlackMessage:
        message = super().to_slack()
        message.fields.append(SlackTextField(heading="Key", text=self.key))
        message.blocks.append(SlackCodeBlock(heading="Error", code=self.error))
        return message
```
Then leave the error out of the main exception message. (In the backtrace it will still be captured because the underlying exception will also be logged directly.)

(Maybe this is overkill; if it doesn't sound like a good idea to you, feel free to drop the idea.)